### DISCUSSION TITLE
why does declarative_base() in 2.0 not work with typing / mypy / vscode / pylance / pycharm?

### DISCUSSION THREAD

**Discussion by toinbis** [sqlalchemy/sqlalchemy#9070] [Usage Questions] [answered]
### Describe the bug

This is how code looks in vscode with sqlalchemy `1.4.46`
![Screenshot 2023-01-08 at 22 08 31](https://user-images.githubusercontent.com/68685/211216796-a4e4f47a-c052-492f-9d09-26dda837c663.png)

This is how code looks in vscode with sqlalchemy `2.0.0rc1`
![Screenshot 2023-01-08 at 22 07 06](https://user-images.githubusercontent.com/68685/211216786-ce76f710-e1d8-40b0-9dad-e27b0b6fdd16.png)

### To Reproduce

```python
Tried to recreate in an empty file, such three lines five the same effect, "metadata" and "create_all" does not bear type hint info for editor.

from sqlalchemy.orm import declarative_base
base = declarative_base()
base.metadata.create_all()
```

### Error

"metadata: Any", "create_all: Any" -> this is all vscode recognizes.

### Versions

- OS: osx
- Python: tested on 3.10 and 3.11
- SQLAlchemy: 2.0.0rc1
- Database: -
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): -

**Comment by zzzeek** [comment, marked as answer]
hey there -

SQLAlchemy 2.0 does not use plugins or stubs for typing.   In 2.0, the canonical way to use a fully typed declarative base is by using the `DeclarativeBase` subclass:

```py

from sqlalchemy.orm import DeclarativeBase

class Base(DeclarativeBase):
   ...

```

this is described in the [upgrade guide](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#step-one-orm-declarative-base-is-superseded-by-orm-declarativebase) for those moving to fully 2.0 style.


The very specific thing you illustrate above, that vscode seems to know about "Base.metadata" in 1.4, I'm not actually sure how that's possible.    The stubs we support for 1.4 are sqlalchemy2-stubs, and the return value of `declarative_base()` in those stubs is simply "type":

https://github.com/sqlalchemy/sqlalchemy2-stubs/blob/86350718c1bde3a8c6f1e83923b20ae00cc76952/sqlalchemy-stubs/orm/decl_api.pyi#L44-L55

whereas in 2.0 at the moment it's `Any`.  

pep-484 typing does not support functions that return classes where those classes would be meaningful as a base class, so 1.4's typing only supported `declarative_base()` when the [mypy plugin were used](https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html#usage), which doesn't work with vscode.      So overall there was never any supported behavior for `declarative_base()` without the mypy plugin being used, I'm not sure what actual stubs your vscode is using with your 1.4 installation.

**Comment by toinbis** [comment]
@zzzeek  - thank you so much for the very clear answer. And also for creating and maintaining sqlalchemy!
